packages/design-system/src/components/primitives/dnd/use-drop.ts

```diff
@@ -0,0 +1,293 @@
+import { useRef, useMemo } from "react";
+import {
+  isEqualRect,
+  isNearEdge,
+  getClosestRectIndex,
+  getPlacementBetween,
+  getPlacementInside,
+  getPlacementNextTo,
+  getIndexAdjustment,
+  type Rect,
+  type Placement,
+} from "./geometry-utils";
+import {
+  type UsableElement,
+  toUseableElement,
+  getLocalChildrenOrientation,
+  getChildrenRects,
+} from "./dom-utils";
+
+// Partial information about a drop target
+// used during the selection of a new drop target
+export type PartialDropTarget<Data> = {
+  data: Data;
+  element: UsableElement;
+};
+
+export type DropTarget<Data> = PartialDropTarget<Data> & {
+  rect: DOMRect;
+  indexWithinChildren: number;
+  placement: Placement;
+};
+
+// We pass around data, to avoid extra data lookups.
+// For example, data found in isDropTarget
+// doesn't have to be looked up again in swapDropTarget.
+export type UseDropProps<Data> = {
+  // To check that the element can qualify as a target
+  isDropTarget: (target: UsableElement) => Data | false;
+
+  // Distance from an edge to set nearEdge to true in swapDropTarget
+  edgeDistanceThreshold?: number;
+
+  // Given the potential target that has passed the isDropTarget check,
+  // and the position of the pointer on the target,
+  // you can swap to another target
+  swapDropTarget: (
+    // undefined is passed when no suitable element is found under the pointer
+    dropTarget: (PartialDropTarget<Data> & { nearEdge: boolean }) | undefined
+  ) => PartialDropTarget<Data>;
+
+  onDropTargetChange: (dropTarget: DropTarget<Data>) => void;
+};
+
+export type UseDropHandlers = {
+  handleMove: (pointerCoordinates: { x: number; y: number }) => void;
+  handleScroll: () => void;
+  handleEnd: () => void;
+  rootRef: (target: Element | null) => void;
+};
+
+const getInitialState = <Data>() => {
+  return {
+    pointerCoordinates: undefined as { x: number; y: number } | undefined,
+    dropTarget: undefined as DropTarget<Data> | undefined,
+    childrenRectsCache: new WeakMap<UsableElement, Rect[]>(),
+    lastCandidateElement: undefined as UsableElement | undefined,
+    lastCandidateIsNearEdge: undefined as boolean | undefined,
+  };
+};
+
+export const useDrop = <Data>(props: UseDropProps<Data>): UseDropHandlers => {
+  // We want to use fresh props every time we use them,
+  // but we don't need to react to updates.
+  // So we can put them in a ref and make useMemo below very efficient.
+  const latestProps = useRef<UseDropProps<Data>>(props);
+  latestProps.current = props;
+
+  const rootRef = useRef<UsableElement>();
```

Comment:
Is UsableElement in the end always an instance element? Mybe we call it InstanceElement then?

Reply:
or WsElement

Reply:
Here it's used for `rootRef` and we don't assume that the root itself is necessarily a drop-target. So, no it doesn't necessarily corresponds to an instance.

Also, it probably would be wrong to use the term Instance here in the dnd hooks?

Reply:
what is rootRef? not a root instance?

Reply:
we are migrating towards having body also as an instance, so I am not sure how you can have anything that isn't an instance

Reply:
true, design system knows nothing about instances ...

Reply:
well lets keep it, I don't have a better name, but Usable sounds weird because usable for what? There should be a better name for it, what is actually an example of an unusable element?

Reply:
I'll try to `as HTMLElement` instead of the helper